### PR TITLE
Add fileReadBytes external

### PR DESCRIPTION
**Pull request by didrikmunther**
Adds the ability to read X amount of characters from a `ReadChannel`. Commonly required in protocols such as JSON-RPC to read the required amount of characters as specified by a `Content-Length: X` header.

### CODE REVIEW

File: stdlib/ext/file-ext.mc [hunk 1]
```diff
@@ -54,6 +54,12 @@ external readLine ! : ReadChannel -> (String, Bool)
 let readLine : ReadChannel -> Option String =
   lam rc. match readLine rc with (s, false) then Some s else None ()
 
+-- Reads a given number of bytes from the file.
```
Comment (marten-voorberg) [inline stdlib/ext/file-ext.mc:62]: What happens if in situations where I read more bytes than that are left? E.g., there are 5 bytes left and I attempt to read 10?

File: stdlib/ext/file-ext.mc [hunk 2]
```diff
@@ -105,6 +111,18 @@ utest
   else ("Error reading file","","","")
 with ("Hello", "Next string", "Final", "EOF") in
 
+-- Test reading x amount of characters from the file
```
Comment (marten-voorberg) [inline stdlib/ext/file-ext.mc:126]: Consider adding a test case for the case in which `None ()` is returned

File: stdlib/ext/file-ext.mc [hunk 3]
```diff
@@ -105,6 +111,18 @@ utest
...
+utest
+  match readOpen filename with Some rc then
+    let l1 = match readBytes rc 3 with Some s then s else "" in
```
Comment (marten-voorberg) [inline stdlib/ext/file-ext.mc]: It seems slightly weird to default to "". 

Maybe `match readBytes rc 3 with Some s in s` 

This might be preferable since not being able to read bytes seems to be an error which defaulting to "" handles silently

File: stdlib/ext/file-ext.mc [hunk 4]
```diff
@@ -105,6 +111,18 @@ utest
...
+    let l2 = match readBytes rc 4 with Some s then s else "" in
+    let l3 = match readBytes rc 0 with Some s then s else "" in
+    let l4 = match readBytes rc 1 with Some s then s else "" in
+    readClose rc;
+    (l1,l2,l3,l4)
+  else ("Error reading file","","","")
```
Comment (marten-voorberg) [inline stdlib/ext/file-ext.mc]: Maybe just throw an `error` instead? E.g. `else error "..."`

**Review by marten-voorberg — CHANGES_REQUESTED** [review]
Looks good! Glad you figured it out. 

I added a couple comments covering some very minor changes.

File: stdlib/ext/file-ext.mc (hunk 3 above)
Comment (didrikmunther) [reply stdlib/ext/file-ext.mc]: That's a good suggestion, I've added it!

File: stdlib/ext/file-ext.mc (hunk 4 above)
Comment (didrikmunther) [reply stdlib/ext/file-ext.mc]: It's written this way in the `readLine` function above, but I think this solution is much cleaner. Is there any downside to throwing an `error` in a `utest` though?

File: stdlib/ext/file-ext.mc (hunk 2 above)
Comment (didrikmunther) [reply stdlib/ext/file-ext.mc:126]: Fixed!

File: stdlib/ext/file-ext.mc (hunk 1 above)
Comment (didrikmunther) [reply stdlib/ext/file-ext.mc:62]: The solution to this turned out to be a completely different implementation in the OCaml code. I have added it now!

Comment (didrikmunther) [reply stdlib/ext/file-ext.mc:62]: The behavior is that it will return `Some a` where `len(a) == 5`. All calls after will return `None ()`

Comment (marten-voorberg) [reply stdlib/ext/file-ext.mc:62]: I think the comment could be a bit clearer on this. E.g.
```
Read the provided number of bytes from a `ReadChannel`.
If there is no more content, `None` is returned.
If the number of remaining bytes is smaller than `len`, 
all remaining bytes are returned

```

Comment (marten-voorberg) [reply stdlib/ext/file-ext.mc:62]: Also, what happens if some error occurs when reading? Is `None` returned or does this raise an error?

File: stdlib/ext/file-ext.mc [hunk 5]
```diff
@@ -105,6 +114,21 @@ utest
   else ("Error reading file","","","")
 with ("Hello", "Next string", "Final", "EOF") in
 
+-- Test reading x amount of characters from the file
+utest
```
Comment (marten-voorberg) [inline stdlib/ext/file-ext.mc:127]: Consider nesting the test cases like this:
```
  match readOpen filename with Some rc then
  	utest readBytes rc 3 with Some "Hel" using optionEq eqString in
  	utest readBytes rc 4 with Some "lo\n" using optionEq eqString in 
  	utest readBytes rc 0 with Some "" using optionEq eqString in 
  	utest readBytes rc 1 with Some "e" using optionEq eqString in 
        -- If there are fewer bytes remaining than we requested, all remaining bytes
       -- are returned.
  	utest readBytes rc 1000 with Some "xt string \nFinal" using optionEq eqString in
  	utest readBytes rc 1000 with None () using optionEq eqString in 
  	utest readBytes rc 1000 with None () using optionEq eqString in 
  	()
  else 
  	error "File could not be read in tests for readBytes"
```

File: stdlib/ext/file-ext.mc (hunk 4 above)
Comment (marten-voorberg) [reply stdlib/ext/file-ext.mc]: Well, I feel the utest should fail normally if the tested function behaves incorrectly (i.e. it does not crash but just returns the wrong value).

However, when there is a problem with the test case itself (e.g., the file can not be opened) this should raise an error IMO. This is because this is not a bug in the tested function, but in the test suite itself.

File: stdlib/ext/file-ext.mc (hunk 5 above)
Comment (didrikmunther) [reply stdlib/ext/file-ext.mc:127]: That looks better, thanks!

File: stdlib/ext/file-ext.mc (hunk 4 above)
Comment (didrikmunther) [reply stdlib/ext/file-ext.mc]: I agree!

File: stdlib/ext/file-ext.mc (hunk 1 above)
Comment (didrikmunther) [reply stdlib/ext/file-ext.mc:62]: I've changed it so that it returns `None` if there was an error, or there was an EOF and empty string returned.

File: stdlib/ext/file-ext.mc [hunk 6]
```diff
@@ -54,6 +56,36 @@ external readLine ! : ReadChannel -> (String, Bool)
...
+    -- tuple: (Content, reached EOF, had error)
+    case ("", true, _) then None () -- EOF
+    case (s, _, false) then Some s -- Success
+    case (_, _, _) then None () -- Error
+  end
+
+-- returns Option content if the requested number of bytes could be read
+-- otherwise, None is returned
+recursive
+  let readBytesBuffered : ReadChannel -> Int -> Option String =
```
Comment (didrikmunther) [inline stdlib/ext/file-ext.mc:76]: I think this could be removed from `stdlib`.
